runtime/compiler/compile/J9SymbolReferenceTable.cpp

```diff
@@ -842,6 +845,86 @@ J9::SymbolReferenceTable::findOrFabricateShadowSymbol(
    return symRef;
    }
 
+TR::SymbolReference *
+J9::SymbolReferenceTable::findFlattenedArrayElementFieldShadow(
+   FlattenedArrayElementFieldShadowKey key,
+   bool isVolatile,
+   bool isPrivate,
+   bool isFinal)
+   {
+   const auto entry = _flattenedArrayElementFieldShadows.find(key);
+   if (entry == _flattenedArrayElementFieldShadows.end())
+      return NULL;
+
+   TR::SymbolReference *symRef = entry->second;
+   int32_t refNum = symRef->getReferenceNumber();
+   TR::Symbol *sym = symRef->getSymbol();
+
+   TR_ASSERT_FATAL(sym->isVolatile() || !isVolatile, "expecting volatile symref but found non-volatile symref #%d\n", refNum);
+   TR_ASSERT_FATAL(!sym->isPrivate() || isPrivate, "expecting non-private symref but found private symref #%d\n", refNum);
+   TR_ASSERT_FATAL(!sym->isFinal() || isFinal, "expecting non-final symref but found final symref #%d\n", refNum);
+
+   return symRef;
+   }
+
+TR::SymbolReference *
+J9::SymbolReferenceTable::findOrFabricateFlattenedArrayShadowSymbol(
+   TR_OpaqueClassBlock *arrayComponentClass,
+   TR::DataType type,
+   uint32_t fieldOffset,
+   bool isVolatile,
+   bool isPrivate,
+   bool isFinal,
+   const char *fieldName,
+   const char *fieldSignature)
+   {
+   int32_t flattenedFieldOffset = (int32_t)fieldOffset - (int32_t)TR::Compiler->om.objectHeaderSizeInBytes();
+   TR_ASSERT_FATAL(flattenedFieldOffset >= 0, "flattenedFieldOffset %d is invalid: fieldOffset %u objectHeaderSizeInBytes %u", fieldOffset, TR::Compiler->om.objectHeaderSizeInBytes());
+
+   FlattenedArrayElementFieldShadowKey key(arrayComponentClass, flattenedFieldOffset, type);
```

Comment:
I'm thinking about the case of a value type that has no fields again.  If two such fields are adjacent to one another in another value type, they might both have the same field offset, which would result in them having equal keys.  So, I'm thinking that `findOrFabricateFlattenedArrayShadowSymbol` should only ever be called for "leaf" fields, never for any "intermediate" value type fields.

If my thinking is correct, would it be possible to add an fatal assertion guarding against that condition?  I'm not exactly sure how that would be checked. . . .

Reply:
`findOrFabricateFlattenedArrayShadowSymbol` is called only if the array element has at least one field. And it is not called on "transient" or "intermediate" fields because  `typeLayout` constructs the flattened "leaf" fields which are used to call `findOrFabricateFlattenedArrayShadowSymbol`.

> If two such fields are adjacent to one another in another value type, they might both have the same field offset, which would result in them having equal keys. 

Sorry I'm not sure I follow. Could you elaborate it?

Reply:
> `findOrFabricateFlattenedArrayShadowSymbol` is called only if the array element has at least one field. And it is not called on "transient" or "intermediate" fields because `typeLayout` constructs the flattened "leaf" fields which are used to call `findOrFabricateFlattenedArrayShadowSymbol`.

:+1:

Reply:
I was imagining a case like the following.  Suppose we had a value type class, `Outer`, with two fields of value types `Inner1` and `Inner2`.  And suppose that all these things are flattened in array elements of type `Outer`.

```
public primitive class Inner1 {}
public primitive class Inner2 {}
public primitive class Outer {
  public final Inner1 v1;
  public final Inner2 v2;
}
...
Outer[] arr = new Outer[10];
```

If we needed to create flattened array element field shadows for the two `v` fields, they would both have keys constructed as `key(QOuter; /* Array component class Outer */, 0 /* offset of zero for both */, TR::Address /* Type */)`, so the keys would be equal.  If the JIT will never need to create shadows for things that are flattened - the fields `v1` and `v2`, in this case - then everything is good, but an assertion guarding against the possibility might be desirable.

Reply:
I don't think there's anything that could be done with a field shadow representing `v1` or `v2` in your example, since there's no data there to access

I don't know what the `type` would be for such a shadow either (whether or not `Inner` is zero-sized). It's `Address` in the example, but there's no reference stored in memory. A flattened field containing a reference could happen to have that reference at the beginning of its layout, but then the data there would be the value of the inner field itself, not the value that contains the inner field

I guess somebody could try to use `Aggregate`? I'm not clear myself on why we still have it

Reply:
This function also doesn't really know about the layout of the type. It's the caller that will know that. I think that to add an assertion checking that it's a leaf field, we would have to add a parameter for the caller to specify, which isn't very useful, since the caller would just pass a constant indicating that it's a leaf field

Reply:
> If we needed to create flattened array element field shadows for the two v fields

`findOrFabricateFlattenedArrayShadowSymbol` will not be called to create flattened array element field shadows for the two `v` fields because after it's flattened, there is essentially zero field for `Outer`.  After it's flattened, `fieldCount` is zero,  no field entry is created. `findOrFabricateFlattenedArrayShadowSymbol` is not called.  As Devin mentioned, only the caller to `findOrFabricateFlattenedArrayShadowSymbol` knows the layout of the type.

```
   const TR::TypeLayout *fieldTypeLayout = comp()->typeLayout(arrayComponentClass);
   size_t fieldCount = fieldTypeLayout->count();

   for (size_t idx = 0; idx < fieldCount; idx++)
      {
      const TR::TypeLayoutEntry &fieldEntry = fieldTypeLayout->entry(idx);
      ... 
      auto * fieldSymRef = comp()->getSymRefTab()->findOrFabricateFlattenedArrayShadowSymbol(arrayComponentClass,
                                                                  fieldEntry._datatype,
                                                                  fieldEntry._offset,
                                                                  fieldEntry._isVolatile,
                                                                  fieldEntry._isPrivate,
                                                                  fieldEntry._isFinal,
                                                                  fieldEntry._fieldname,
                                                                  fieldEntry._typeSignature);
      } 
```

Reply:
@hzongaro I'm wondering if the above comments address your questions. Meanwhile I've created a draft PR (https://github.com/eclipse-openj9/openj9/pull/14098) on the transformation in VP which will also help answer the question on zero field situation.

Reply:
Yes, Devin is correct that this method doesn't have enough information to apply such an assertion.  I think I was being a bit too paranoid about the possibility that somebody might come along in a couple of years and make a change that could result in `findOrFabricateFlattenedArrayShadowSymbol` being called for one of the "intermediate" flattened fields.

You can mark this as resolved.